File: tests/widgets/controls/CheckBox.m
#include <Cocoa/Cocoa.h>

static const char* nsstr(NSString* s) { return s ? [s UTF8String] : ""; }

@interface Window : NSWindow {
  NSButton* checkBox1;
  NSButton* checkBox2;
  NSButton* checkBox3;
  NSButton* checkBox4;
  NSButton* checkBox5;
}
- (instancetype)init;
- (BOOL)windowShouldClose:(id)sender;
- (IBAction) OnCheckBox1Click:(id)sender;
- (IBAction) OnCheckBox2Click:(id)sender;
- (IBAction) OnCheckBox3Click:(id)sender;
- (IBAction) OnCheckBox4Click:(id)sender;
- (IBAction) OnCheckBox5Click:(id)sender;
- (NSString*) stateToString:(NSControlStateValue)state;
- (void)dumpLayout:(NSString*)stage;
@end

@implementation Window
- (void)dumpView:(NSView*)view name:(NSString*)name {
  NSRect frame = [view frame];
  NSRect bounds = [view bounds];
  NSLog(@"[%@] frame=(%.1f,%.1f %.1fx%.1f) bounds=(%.1f,%.1f %.1fx%.1f) autoresizeMask=0x%lx",
      name,
      frame.origin.x, frame.origin.y, frame.size.width, frame.size.height,
      bounds.origin.x, bounds.origin.y, bounds.size.width, bounds.size.height,
      (unsigned long)[view autoresizingMask]);
  printf("[%s] frame=(%.1f,%.1f %.1fx%.1f) bounds=(%.1f,%.1f %.1fx%.1f) autoresizeMask=0x%lx\n",
      nsstr(name),
      frame.origin.x, frame.origin.y, frame.size.width, frame.size.height,
      bounds.origin.x, bounds.origin.y, bounds.size.width, bounds.size.height,
      (unsigned long)[view autoresizingMask]);
}

- (void)dumpLayout:(NSString*)stage {
  NSRect windowFrame = [self frame];
  NSRect contentRect = [self contentRectForFrameRect:windowFrame];
  NSLog(@"[Window %@] frame=(%.1f,%.1f %.1fx%.1f) contentRect=(%.1f,%.1f %.1fx%.1f)",
      stage,
      windowFrame.origin.x, windowFrame.origin.y, windowFrame.size.width, windowFrame.size.height,
      contentRect.origin.x, contentRect.origin.y, contentRect.size.width, contentRect.size.height);
  printf("[Window %s] frame=(%.1f,%.1f %.1fx%.1f) contentRect=(%.1f,%.1f %.1fx%.1f)\n",
      nsstr(stage),
      windowFrame.origin.x, windowFrame.origin.y, windowFrame.size.width, windowFrame.size.height,
      contentRect.origin.x, contentRect.origin.y, contentRect.size.width, contentRect.size.height);
  [self dumpView:[self contentView] name:@"contentView"];
  [self dumpView:checkBox1 name:@"checkBox1"];
  [self dumpView:checkBox2 name:@"checkBox2"];
  [self dumpView:checkBox3 name:@"checkBox3"];
  [self dumpView:checkBox4 name:@"checkBox4"];
  [self dumpView:checkBox5 name:@"checkBox5"];
  NSLog(@"[checkBox1] state=%ld mixed=%d bezel=%ld alignment=%ld title='%@'",
      (long)[checkBox1 state], [checkBox1 allowsMixedState], (long)[checkBox1 bezelStyle], (long)[checkBox1 alignment], [checkBox1 title]);
  NSLog(@"[checkBox2] state=%ld mixed=%d bezel=%ld alignment=%ld title='%@'",
      (long)[checkBox2 state], [checkBox2 allowsMixedState], (long)[checkBox2 bezelStyle], (long)[checkBox2 alignment], [checkBox2 title]);
  NSLog(@"[checkBox3] state=%ld mixed=%d bezel=%ld alignment=%ld title='%@'",
      (long)[checkBox3 state], [checkBox3 allowsMixedState], (long)[checkBox3 bezelStyle], (long)[checkBox3 alignment], [checkBox3 title]);
  NSLog(@"[checkBox4] state=%ld mixed=%d bezel=%ld alignment=%ld title='%@'",
      (long)[checkBox4 state], [checkBox4 allowsMixedState], (long)[checkBox4 bezelStyle], (long)[checkBox4 alignment], [checkBox4 title]);
  NSLog(@"[checkBox5] state=%ld mixed=%d bezel=%ld alignment=%ld title='%@'",
      (long)[checkBox5 state], [checkBox5 allowsMixedState], (long)[checkBox5 bezelStyle], (long)[checkBox5 alignment], [checkBox5 title]);
}

- (instancetype)init {
  checkBox1 = [[[NSButton alloc] initWithFrame:NSMakeRect(30, 250, 105, 20)] autorelease];
  [checkBox1 setTitle:@"Unchecked"];
  [checkBox1 setButtonType:NSButtonTypeSwitch];
  [checkBox1 setTarget:self];
  [checkBox1 setAction:@selector(OnCheckBox1Click:)];
  [checkBox1 setAutoresizingMask:NSViewMaxXMargin | NSViewMinYMargin];
  [checkBox1 setState:NSControlStateValueOff];
  
  checkBox2 = [[[NSButton alloc] initWithFrame:NSMakeRect(30, 220, 105, 20)] autorelease];
  [checkBox2 setTitle:@"Checked"];
  [checkBox2 setButtonType:NSButtonTypeSwitch];
  [checkBox2 setTarget:self];
  [checkBox2 setAction:@selector(OnCheckBox2Click:)];
  [checkBox2 setAutoresizingMask:NSViewMaxXMargin | NSViewMinYMargin];
  [checkBox2 setState:NSControlStateValueOn];
  
  checkBox3 = [[[NSButton alloc] initWithFrame:NSMakeRect(30, 190, 105, 20)] autorelease];
  [checkBox3 setTitle:@"Mixed"];
  [checkBox3 setAllowsMixedState:YES];
  [checkBox3 setButtonType:NSButtonTypeSwitch];
  [checkBox3 setTarget:self];
  [checkBox3 setAction:@selector(OnCheckBox3Click:)];
  [checkBox3 setAutoresizingMask:NSViewMaxXMargin | NSViewMinYMargin];
  [checkBox3 setState:NSControlStateValueMixed];

  checkBox4 = [[[NSButton alloc] initWithFrame:NSMakeRect(30, 160, 105, 25)] autorelease];
  [checkBox4 setTitle:@"Checked"];
  [checkBox4 setButtonType:NSButtonTypeOnOff];
  [checkBox4 setBezelStyle:NSBezelStyleRounded];
  [checkBox4 setTarget:self];
  [checkBox4 setAction:@selector(OnCheckBox4Click:)];
  [checkBox4 setAutoresizingMask:NSViewMaxXMargin | NSViewMinYMargin];
  [checkBox4 setState:NSControlStateValueOn];

  checkBox5 = [[[NSButton alloc] initWithFrame:NSMakeRect(30, 130, 105, 25)] autorelease];
  [checkBox5 setTitle:@"Unchecked"];
  [checkBox5 setButtonType:NSButtonTypeOnOff];
  [checkBox5 setBezelStyle:NSBezelStyleRounded];
  [checkBox5 setTarget:self];
  [checkBox5 setAction:@selector(OnCheckBox5Click:)];
  [checkBox5 setAutoresizingMask:NSViewMaxXMargin | NSViewMinYMargin];
  [checkBox5 setState:NSControlStateValueOff];

  [super initWithContentRect:NSMakeRect(100, 100, 300, 300) styleMask:NSWindowStyleMaskTitled | NSWindowStyleMaskClosable | NSWindowStyleMaskMiniaturizable | NSWindowStyleMaskResizable backing:NSBackingStoreBuffered defer:NO];
  [self setTitle:@"CheckBox example"];
  [[self contentView] addSubview:checkBox1];
  [[self contentView] addSubview:checkBox2];
  [[self contentView] addSubview:checkBox3];
  [[self contentView] addSubview:checkBox4];
  [[self contentView] addSubview:checkBox5];
  [self setIsVisible:YES];
  [self dumpLayout:@"init"];
  return self;
}

- (BOOL)windowShouldClose:(id)sender {
  [NSApp terminate:sender];
  return YES;
}

- (IBAction) OnCheckBox1Click:(id)sender {
  [checkBox1 setState:NSControlStateValueOff];
  [checkBox1 setTitle: [self stateToString: [checkBox1 state]]];
  [self dumpLayout:@"checkBox1-click"];
}

- (IBAction) OnCheckBox2Click:(id)sender {
  [checkBox2 setTitle: [self stateToString: [checkBox2 state]]];
  [self dumpLayout:@"checkBox2-click"];
}

- (IBAction) OnCheckBox3Click:(id)sender {
  [checkBox3 setTitle: [self stateToString: [checkBox3 state]]];
  [self dumpLayout:@"checkBox3-click"];
}

- (IBAction) OnCheckBox4Click:(id)sender {
  [checkBox4 setTitle: [self stateToString: [checkBox4 state]]];
  [self dumpLayout:@"checkBox4-click"];
}

- (IBAction) OnCheckBox5Click:(id)sender {
  [checkBox5 setState:NSControlStateValueOff];
  [checkBox5 setTitle: [self stateToString: [checkBox5 state]]];
  [self dumpLayout:@"checkBox5-click"];
}

- (NSString*) stateToString:(NSControlStateValue)state {
  switch (state) {
    case NSControlStateValueOff: return @"Unchecked";
    case NSControlStateValueOn: return @"Checked";
    case NSControlStateValueMixed: return @"Mixed";
  }
  return @"Unchecked";
}
@end

int main(int argc, char *argv[]) {
    [NSApplication sharedApplication];
    [NSApp setActivationPolicy:NSApplicationActivationPolicyRegular];

    Window *window = [[Window alloc] init];
    [window makeKeyAndOrderFront:nil];

    [NSApp activateIgnoringOtherApps:YES];
    if (getenv("CHECKBOX_DUMP_LAYOUT_ONCE") != NULL) {
      [NSApp performSelector:@selector(terminate:) withObject:nil afterDelay:0.2];
    }
    [NSApp run];
    return 0;
}

File: tests/widgets/controls/checkbox.nim
import std/[os, strutils]

import knutella/appkit
import knutella/objc

proc maxFramesFromEnv(defaultValue = -1): int =
  let raw = getEnv("KNUTELLA_EXAMPLE_FRAMES").strip()
  if raw.len == 0:
    return defaultValue
  try:
    parseInt(raw)
  except ValueError:
    defaultValue

proc debugRenderDumpEnabled(): bool =
  getEnv("KNUTELLA_APPKIT_DEBUG_RENDER").strip().toLowerAscii() in
    ["1", "true", "yes", "on"]

proc format1(value: float32): string =
  formatFloat(value.float, ffDecimal, 1)

proc formatRect(rect: NSRect): string =
  "(" & format1(rect.origin.x) & "," & format1(rect.origin.y) & " " &
    format1(rect.size.width) & "x" & format1(rect.size.height) & ")"

proc intBool(flag: bool): int =
  if flag: 1 else: 0

proc dumpViewLine(view: NSView, name: string) =
  if view.isNil:
    return
  echo "[",
    name,
    "] frame=",
    formatRect(view.frame()),
    " bounds=",
    formatRect(view.bounds()),
    " autoresizeMask=0x",
    toHex(view.autoresizingMask(), 2)

proc dumpCheckBoxLayout(
    stage: string,
    window: NSWindow,
    checkBox1: NSButton,
    checkBox2: NSButton,
    checkBox3: NSButton,
    checkBox4: NSButton,
    checkBox5: NSButton,
) =
  if window.isNil:
    return
  let windowFrame = window.frame()
  let contentRect = window.contentRectForFrameRect(windowFrame)
  let contentView = window.contentView()
  echo "[Window ",
    stage, "] frame=", formatRect(windowFrame), " contentRect=", formatRect(contentRect)
  dumpViewLine(contentView, "contentView")
  dumpViewLine(checkBox1), "checkBox1" as NSView
  dumpViewLine(checkBox2), "checkBox2" as NSView
  dumpViewLine(checkBox3), "checkBox3" as NSView
  dumpViewLine(checkBox4), "checkBox4" as NSView
  dumpViewLine(checkBox5), "checkBox5" as NSView
  echo "[checkBox1] state=",
    checkBox1.state(),
    " mixed=",
    intBool(checkBox1.allowsMixedState()),
    " alignment=",
    checkBox1.alignment(),
    " title='",
    checkBox1.title(),
    "'"
  echo "[checkBox2] state=",
    checkBox2.state(),
    " mixed=",
    intBool(checkBox2.allowsMixedState()),
    " alignment=",
    checkBox2.alignment(),
    " title='",
    checkBox2.title(),
    "'"
  echo "[checkBox3] state=",
    checkBox3.state(),
    " mixed=",
    intBool(checkBox3.allowsMixedState()),
    " alignment=",
    checkBox3.alignment(),
    " title='",
    checkBox3.title(),
    "'"
  echo "[checkBox4] state=",
    checkBox4.state(),
    " mixed=",
    intBool(checkBox4.allowsMixedState()),
    " alignment=",
    checkBox4.alignment(),
    " title='",
    checkBox4.title(),
    "'"
  echo "[checkBox5] state=",
    checkBox5.state(),
    " mixed=",
    intBool(checkBox5.allowsMixedState()),
    " alignment=",
    checkBox5.alignment(),
    " title='",
    checkBox5.title(),
    "'"
  if debugRenderDumpEnabled():
    debugDumpWindowRenderTree(window)

objcImpl:
  type CheckBoxWindow = object of NSWindow
    xCheckBox1: NSButton
    xCheckBox2: NSButton
    xCheckBox3: NSButton
    xCheckBox4: NSButton
    xCheckBox5: NSButton

  method init*(self: var CheckBoxWindow): CheckBoxWindow =
    var base = self.initWithContentRect(
      100.0,
      100.0,
      300.0,
      328.0,
      NSTitledWindowMask or NSClosableWindowMask or NSMiniaturizableWindowMask or
        NSResizableWindowMask,
      NSBackingStoreBuffered,
      false,
    )
    result = asTypeRaw[CheckBoxWindow](base.value)
    base.value = nil
    if result.isNil:
      return

    var contentAlloc = NSView.alloc()
    var contentView = contentAlloc.initWithFrame(0.0, 0.0, 300.0, 300.0)
    contentAlloc.value = nil
    result.setContentView(contentView)

    var checkBox1Alloc = NSButton.alloc()
    result.xCheckBox1 = checkBox1Alloc.initWithFrame(30.0, 250.0, 105.0, 20.0)
    checkBox1Alloc.value = nil
    result.xCheckBox1.setTitle(@ns"Unchecked")
    result.xCheckBox1.setButtonType(NSSwitchButton.cint)
    result.xCheckBox1.setTarget(ID(value: result.value))
    result.xCheckBox1.setAction(getSelector("OnCheckBox1Click:"))
    result.xCheckBox1.setAlignment(NSNaturalTextAlignment)
    result.xCheckBox1.setAutoresizingMask(NSViewMaxXMargin or NSViewMinYMargin)
    result.xCheckBox1.setState(NSOffState.cint)

    var checkBox2Alloc = NSButton.alloc()
    result.xCheckBox2 = checkBox2Alloc.initWithFrame(30.0, 220.0, 105.0, 20.0)
    checkBox2Alloc.value = nil
    result.xCheckBox2.setTitle(@ns"Checked")
    result.xCheckBox2.setButtonType(NSSwitchButton.cint)
    result.xCheckBox2.setTarget(ID(value: result.value))
    result.xCheckBox2.setAction(getSelector("OnCheckBox2Click:"))
    result.xCheckBox2.setAlignment(NSNaturalTextAlignment)
    result.xCheckBox2.setAutoresizingMask(NSViewMaxXMargin or NSViewMinYMargin)
    result.xCheckBox2.setState(NSOnState.cint)

    var checkBox3Alloc = NSButton.alloc()
    result.xCheckBox3 = checkBox3Alloc.initWithFrame(30.0, 190.0, 105.0, 20.0)
    checkBox3Alloc.value = nil
    result.xCheckBox3.setTitle(@ns"Mixed")
    result.xCheckBox3.setAllowsMixedState(true)
    result.xCheckBox3.setButtonType(NSSwitchButton.cint)
    result.xCheckBox3.setTarget(ID(value: result.value))
    result.xCheckBox3.setAction(getSelector("OnCheckBox3Click:"))
    result.xCheckBox3.setAlignment(NSNaturalTextAlignment)
    result.xCheckBox3.setAutoresizingMask(NSViewMaxXMargin or NSViewMinYMargin)
    result.xCheckBox3.setState(NSMixedState.cint)

    var checkBox4Alloc = NSButton.alloc()
    result.xCheckBox4 = checkBox4Alloc.initWithFrame(30.0, 160.0, 105.0, 25.0)
    checkBox4Alloc.value = nil
    result.xCheckBox4.setTitle(@ns"Checked")
    result.xCheckBox4.setButtonType(NSOnOffButton.cint)
    result.xCheckBox4.setBezelStyle(NSRoundedBezelStyle)
    result.xCheckBox4.setTarget(ID(value: result.value))
    result.xCheckBox4.setAction(getSelector("OnCheckBox4Click:"))
    result.xCheckBox4.setAlignment(NSCenterTextAlignment)
    result.xCheckBox4.setAutoresizingMask(NSViewMaxXMargin or NSViewMinYMargin)
    result.xCheckBox4.setState(NSOnState.cint)

    var checkBox5Alloc = NSButton.alloc()
    result.xCheckBox5 = checkBox5Alloc.initWithFrame(30.0, 130.0, 105.0, 25.0)
    checkBox5Alloc.value = nil
    result.xCheckBox5.setTitle(@ns"Unchecked")
    result.xCheckBox5.setButtonType(NSOnOffButton.cint)
    result.xCheckBox5.setBezelStyle(NSRoundedBezelStyle)
    result.xCheckBox5.setTarget(ID(value: result.value))
    result.xCheckBox5.setAction(getSelector("OnCheckBox5Click:"))
    result.xCheckBox5.setAlignment(NSCenterTextAlignment)
    result.xCheckBox5.setAutoresizingMask(NSViewMaxXMargin or NSViewMinYMargin)
    result.xCheckBox5.setState(NSOffState.cint)

    contentView.addSubview(result.xCheckBox1)
    contentView.addSubview(result.xCheckBox2)
    contentView.addSubview(result.xCheckBox3)
    contentView.addSubview(result.xCheckBox4)
    contentView.addSubview(result.xCheckBox5)

    result.setTitle(@ns"CheckBox example")
    result.setIsVisible(true)
    dumpCheckBoxLayout(
      "init",
      result as NSWindow,
      result.xCheckBox1,
      result.xCheckBox2,
      result.xCheckBox3,
      result.xCheckBox4,
      result.xCheckBox5,
    )
    contentView.value = nil

  method windowShouldClose*(self: CheckBoxWindow, sender: NSObject): bool =
    if sender.isNil:
      NSApp().stop()
      return true
    NSApp().stop()
    true

  method stateToString*(self: CheckBoxWindow, state: int): NSString =
    if self.isNil:
      return @ns"Unchecked"
    case state
    of NSOffState:
      @ns"Unchecked"
    of NSOnState:
      @ns"Checked"
    else:
      @ns"Mixed"

  method OnCheckBox1Click*(self: CheckBoxWindow, sender: NSObject) =
    if sender.isNil:
      return
    self.xCheckBox1.setState(NSOffState.cint)
    self.xCheckBox1.setTitle(self.stateToString(self.xCheckBox1.state()))
    dumpCheckBoxLayout(
      "checkBox1-click",
      self as NSWindow,
      self.xCheckBox1,
      self.xCheckBox2,
      self.xCheckBox3,
      self.xCheckBox4,
      self.xCheckBox5,
    )

  method OnCheckBox2Click*(self: CheckBoxWindow, sender: NSObject) =
    if sender.isNil:
      return
    self.xCheckBox2.setTitle(self.stateToString(self.xCheckBox2.state()))
    dumpCheckBoxLayout(
      "checkBox2-click",
      self as NSWindow,
      self.xCheckBox1,
      self.xCheckBox2,
      self.xCheckBox3,
      self.xCheckBox4,
      self.xCheckBox5,
    )

  method OnCheckBox3Click*(self: CheckBoxWindow, sender: NSObject) =
    if sender.isNil:
      return
    self.xCheckBox3.setTitle(self.stateToString(self.xCheckBox3.state()))
    dumpCheckBoxLayout(
      "checkBox3-click",
      self as NSWindow,
      self.xCheckBox1,
      self.xCheckBox2,
      self.xCheckBox3,
      self.xCheckBox4,
      self.xCheckBox5,
    )

  method OnCheckBox4Click*(self: CheckBoxWindow, sender: NSObject) =
    if sender.isNil:
      return
    self.xCheckBox4.setTitle(self.stateToString(self.xCheckBox4.state()))
    dumpCheckBoxLayout(
      "checkBox4-click",
      self as NSWindow,
      self.xCheckBox1,
      self.xCheckBox2,
      self.xCheckBox3,
      self.xCheckBox4,
      self.xCheckBox5,
    )

  method OnCheckBox5Click*(self: CheckBoxWindow, sender: NSObject) =
    if sender.isNil:
      return
    self.xCheckBox5.setState(NSOffState.cint)
    self.xCheckBox5.setTitle(self.stateToString(self.xCheckBox5.state()))
    dumpCheckBoxLayout(
      "checkBox5-click",
      self as NSWindow,
      self.xCheckBox1,
      self.xCheckBox2,
      self.xCheckBox3,
      self.xCheckBox4,
      self.xCheckBox5,
    )

  method dealloc(self: CheckBoxWindow) {.used.} =
    self.xCheckBox1 = NSButton(value: nil)
    self.xCheckBox2 = NSButton(value: nil)
    self.xCheckBox3 = NSButton(value: nil)
    self.xCheckBox4 = NSButton(value: nil)
    self.xCheckBox5 = NSButton(value: nil)
    destroyIvarFields(self)
    discard callSuperIdFrom(CheckBoxWindow, self, getSelector("dealloc"))

when isMainModule:
  var app = NSApp()
  var windowAlloc = CheckBoxWindow.alloc()
  var window = initOwned(move(windowAlloc))

  app.addWindow(window) as NSWindow
  window.makeKeyAndOrderFront(app) as NSObject

  try:
    let maxFrames = maxFramesFromEnv()
    if maxFrames < 0:
      app.run()
    else:
      discard app.runForFrames(maxFrames)
  except CatchableError as exc:
    echo "Unable to run checkbox example: ", exc.msg

  window.value = nil
  app.value = nil
